### DISCUSSION TITLE
Some issues regarding units

### DISCUSSION THREAD

**Discussion by mpradovelasco** [OpenModelica/OpenModelica#10776] [Q&A] [answered]
Hello,

I have several doubts concerning the units that appear in OMEdit 1.21 and the latest 1.22-dev:

i) For OMEdit latest 1.22-dev. The combo of display and SI units does not appear in the dialog box when "Enable Instance API" is on.

ii) For OMEdit 1.21 and latest 1.22-dev. I get parsing error when using a new quantity / unit that it is not known in Modelica.Units.SI. For example, when modifies a real array with this quantity/unit:
  ```
Signals.Metrics.AUC AUC(soch(each quantity = "MassConcentrationbyTime", each unit = "kg/m3.s", each displayUnit = "ug/l.h"))
```
Does exist any procedure to solve this problem?

iii)  For OMEdit 1.21 and latest 1.22-dev. Related to the ii issue, as I need to use additional units to those of OpenModelica, I am using a complementary .mo file with new units, and I import both (Modelica.Units.SI and the new one). However, I wonder it there is a better solution to perform this addition. Moreover, is it necessary to implement the conversion function between display and SI units in the case of a novel dimensional unit?

Thanks in advance!

**Comment by casella** [comment]
i) I noticed this as well some time ago, but I understand it's been resolved, please try the latest nightly build
ii) and iii) I think we should make it possible for users to customize the unit conversion. This is something we'd like to do in the near future, if you have requirements or suggestions for that, please go ahead and state them here.

**Reply by mpradovelasco** [reply]
Hi @casella, 
I will try to answer to the ii and iii) issues with this message. I have found several units related issues (with "Enable Instance API" turned OFF) that I point in first place, followed by several suggestions/requirements fo unit conversion of new units (quantities), as you indicated. 

**Potential bugs**
  1. The modifiers tab in dialog box (double click on a model component) seems do not work. I have tried this function with several classes and attributes and it gives back always a scripting error. For instance, (trying to modify the start attribute of the Real xs of sAUC component):
  `Unable to set the element modifier value using command  setElementModifierValue (CybSim.test.drF2cmpLivGut, sAUC.xs, $Code(=(start = 1)))`
   I know it is easy to perform this modification through the code view, but it would be much more accessible for users without a good knowledge of Modelica.
  The reason to address here this issue is the interest to apply the modifiers tab to modify the unit, displayUnit, and quantity attributes of a Real variable, which is an output of a block component used to compute metrics and others magnitudes (that depend on the use of the final model). 

  2. For single Real and one-dimension Real arrays variables (parameters):
    a) It starts presenting its default value in SI in the edit box, but the displayUnit does not show the SI unit.
    b) Once the edit box is modified, the displayUnit cannot be modified, and the dialog box needs to be close to change it. 

  3. For double-dimension Real arrays variables (parameters), when trying to use the notation {{}}, the displayUnit box changes to SI, what impedes to use the displayUnit value.
  
  4. The pretty printing algorithm triggers almost always that the displayUnit box is changed. 

**Conversion of new units**

1. I use a package where new Real types with new quantities are defined, like:
```
type MassFlowRatePerArea = Real (
      final quantity="MassFlowRatePerArea", 
      final unit="kg/(s.m2)");
```
Thus, the Modelica.Units.SI and the new one CybSim.SIUnits, must be imported. It could be interested that OpenModelica could detect repeated quantities, to control the case that Modelica.Units.SI would be extended.
This structure works fine. I tried also to inherit the Modelica.Units.SI inside CybSim.SIUnits, to import only the last package, without success.

2. OpenModelica seems work fine with these new types in equations. However, it would be important to enable the conversion of new Real (quantities) to plot them and to manage their input values in dialog boxes.

Thank you for your support!

**Comment by mpradovelasco** [comment]
Hi,

Concerning the i) issue, I have just tested again with the latest 1.22-dev, but the units do not appear in the dialog boxes when "Enable instance API" is ON (Windows 11). 

Concerning the issues ii-iii) I will state here some suggestions as you proposed me in the previous comment. Thank you in advance!

**Comment by casella** [comment]
@mpradovelasco I tested this very simple MWE:
```modelica
model TestParameterUnits
  Modelica.Electrical.Analog.Basic.Resistor resistor annotation(
    Placement(visible = true, transformation(origin = {28, -6}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
annotation(
    uses(Modelica(version = "4.0.0")));
end TestParameterUnits;
```

with OpenModelica v1.22.0-dev-155-g28817ae75c, Enable Instance API on, Windows 10. If I double-click on `resistor`, I get

![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/d74ebfaa-32d2-4c6b-9b48-3cb74cfa1e9e)

which LGTM.

Not sure what is going on with your tests. Could you:
- attach the (zipped) .mo file of a test?
- try resetting Tools | Options to factory settings with the Reset button, then turn on Enable Instance API? Maybe you have some setting that interferes with this feature

**Comment by mpradovelasco** [comment]
Hi @casella, I have tested again this issue in Windows 10, with your simple model, using several Modelica Examples (examples for the MSL, for example from Modelica.Blocks.Examples), and finally using my library with the same result, that is, the units disappear when "Enable Instance API" is ON.

I may paste the dialog box of the Resistor, of your simple model:
![imagen](https://github.com/OpenModelica/OpenModelica/assets/114144098/40e40b90-55a1-4142-916a-804c751bde5e)

The units appear correctly when Enable Instance API is OFF. 

I have performed a reset of Tool | Options to factory settings, turn on the Enable Instance API, close and open OMEdit, and repeat all the tests with the same results. 

The header of the "About OMEdit" says:

Connected to v1.22.0-dev-155-g28817ae75c-cmake
Connected to OMSimulator v2.1.1.post188-gaf996ad-mingw-debug
Installation path C:/DESARROLLOYDATOS/Areas/Repos/OpenModelica/build_cmake/install_cmake

The initial test was performed in a Windows 11 SO (a different machine) with the same OMEdit version and same results. 

Thanks again for your effort

**Comment by casella** [comment]
@adeas31 this is weird. At some point I myself experienced this behaviour, but I interpreted it as a glitch in a nightly build that was later fixed. Maybe it was something else. Any idea?

**Comment by adeas31** [comment]
Yes, that is really weird. I can't reproduce this.

@mpradovelasco please send us `%TEMP%/OpenModelica/OMEdit/omeditcommunication.log`

**Comment by mpradovelasco** [comment]
Hello @adeas31 , I am attaching that file to this message. Previously to get it, I have repeated the test performed with the model of a single resistor in OMEdit (latest 1.22 dev) on the machine with Windows 11, with the same results. The .log file is from the Windows 11 machine.

[omeditcommunication.log](https://github.com/OpenModelica/OpenModelica/files/11650762/omeditcommunication.log)

Thank you

**Comment by adeas31** [comment]
@mpradovelasco the log file is not complete. You need to close OMEdit so that all the information is properly written to the file.

**Comment by mpradovelasco** [comment]
Hi @adeas31, I am attaching again the requested file after closing OMEdit. It refers to the Windows 10 installation, with the Enable Instance API on, and after opening the dialog box of the resistor from the previous single model.

[omeditcommunication.log](https://github.com/OpenModelica/OpenModelica/files/11676107/omeditcommunication.log)

Thanks

**Comment by adeas31** [comment]
The log file looks fine. However, I noticed that you added the test model `TestParameterUnits` inside your library `CybSim`. Can you please test without loading your library? Just start OMEdit without any MSL and load only `TestParameterUnits`.

```
model TestParameterUnits
  Modelica.Electrical.Analog.Basic.Resistor resistor annotation(
    Placement(visible = true, transformation(origin = {28, -6}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
annotation(
    uses(Modelica(version = "4.0.0")));
end TestParameterUnits;
```

**Comment by mpradovelasco** [comment]
Hello again, 
I have repeated the opening the OMedit without CybSim and even MSL. I have open the single model test. and subsequently the dialog box of the resistor with double click. Enable Instance API is ON. The issue persists in my system. In this case, I am attaching the log of my Windows 11 machine. 

[omeditcommunication.log](https://github.com/OpenModelica/OpenModelica/files/11687011/omeditcommunication.log)

Thanks

**Comment by casella** [comment]
@mpradovelasco regarding the first bug you reported, please follow #5862.

**Comment by casella** [comment]
For the other bugs you reported, would you be so kind to open one issue for each, attaching a minimal working example that reproduces the issue? Thanks!

**Comment by mpradovelasco** [comment]
Hi @casella  

I have tested again the fail to present the units in the dialog box of the Resistor of the minimal example (TestParameterUnits) above with the "Enable Instance API" turned ON in the last OMEdit 1.22-dev ( v1.22.0-dev-161-g8eff86a43d) and it is solved. I have confirmed that it works also in CybSim. Great!

I am going to review other pointed issues and will open them as potential bugs if they persist (or commented here the contrary). 

Thank you!

**Reply by mpradovelasco** [reply]
Excuse me. The presentation of units in dialog box fail still persists in latest 1.22-dev. I had the "Enable Instance API" turned OFF. 

I will review the others issues. Thanks.

**Reply by casella** [reply]
Still a mystery why this happens. @adeas31 any clue?

**Reply by adeas31** [reply]
I have no idea. This is very strange. I looked in the logs that @mpradovelasco sent and everything looks fine. Units should be visible regardless of instance API switch.

**Reply by casella** [reply]
Could it be a Win 11 problem? But I understand it is also present with Win 10, right?

**Reply by mpradovelasco** [reply]
Yes, it appears in Win 11 and Win 10. 
I am going to repeat the testing in Win 10 with the latest 1.22-dev, beside trying to modify the graphic setting (and even local region settings) of the equipments, to have more data and inform you. 
If no other clues appear we could plan a brief session of zoom or team and even I could provide you during the session control of my laptop through teamViewer, for example, to allow you the management of openModelica as you need. 
Thanks

**Reply by casella** [reply]
Sound good, please arrange a session with @adeas31

**Reply by mpradovelasco** [reply]
Hi,
After a new review as commented, the problem keeps in Win 10 and 11 with the 1.22-dev (v1.22.0-dev-161-g8eff86a43d). I have changed scales, resolutions, regions and even place of openModelica installation or access with administrator role without success.
I will contact with @adeas31 through email to arrange a session. 
Thanks

**Comment by mpradovelasco** [comment, marked as answer]
Hi again, the trouble of units presentation in dialog boxes is solved in 1.22-dev as commented in #10846. 
Thanks!
